File: 1122. Relative Sort Array/1122.py
class Solution:
    def relativeSortArray(self, arr1, arr2):
        from collections import OrderedDict
        mydict = OrderedDict()
        arr1.sort()
        for i in arr1:
            v = mydict.get(i, 0)
            mydict[i] = v+1
        
        loc = 0
        for i in arr2:
            count = mydict[i]
            while count>0:
                arr1[loc] = i
                count -= 1
                loc += 1
            mydict[i] = 0

        for i in mydict:
            count = mydict[i]
            while count>0:
                arr1[loc] = i
                count -= 1
                loc += 1
            mydict[i] = 0
        return arr1

    # 这个对sorted函数用的挺好的
    # key返回一个元组，两个值来进行比较，也是第一次见到这种操作
    def relativeSortArray1(self, arr1, arr2):
        def custom_sort(val):
            try:
                index = arr2.index(val)
                return (0, index)
            except ValueError:
                return (1, val)
        return sorted(arr1, key=custom_sort)

    # 这个用了很多自带的函数
    def relativeSortArray2(self, arr1, arr2):
        res = []
        for i in arr2:
            res = res + [i] * arr1.count(i)
        return res + sorted(filter(lambda x: arr2.count(x) == 0, arr1))

a = Solution()
print(a.relativeSortArray([28,6,22,8,44,17],[22,28,8,6]))

File: 1122. Relative Sort Array/1122.go
package leetcode

import "sort"

func relativeSortArray(arr1 []int, arr2 []int) []int {
	sort.Ints(arr1)
	mymap := make(map[int]int, len(arr1))
	for _,v := range arr1{
		mapv, ok := mymap[v]
		if !ok{
			mymap[v] = 1
		}else{
			mymap[v] = mapv + 1
		}
	}

	myslice := make([]int, len(arr1))
	var loc int
	for _,v := range arr2{
		mapv, _ := mymap[v]
		for mapv > 0{
			myslice[loc] = v
			mapv --
			loc ++
		}
		mymap[v] = 0
	}

	for _,v := range arr1{
		mapv, _ := mymap[v]
		for mapv > 0{
			myslice[loc] = v
			mapv --
			loc ++
		}
		mymap[v] = 0
	}

	return myslice
}

// 上面那个测试内存调用4次，这个只有一次，有点谜啊
// 这个用数组的方式来代替字典，也见过几次，感觉这个的效果貌似比字典好些
func relativeSortArray1(A, B []int) []int {
	count := [1001]int{}
	for _, a := range A {
		count[a]++
	}

	res := make([]int, 0, len(A))
	for _, b := range B {
		for count[b] > 0 {
			res = append(res, b)
			count[b]--
		}
	}
	for i := 0; i < 1001; i++ {
		for count[i] > 0 {
			res = append(res, i)
			count[i]--
		}
	}

	return res
}